### ISSUE TITLE
BigQuery: Usage of `DECLARE`ed variable triggers L027

### ISSUE THREAD

**Issue by dmohns** [sqlfluff/sqlfluff#2991]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When declaring a BigQuery variable using `DECLARE` and then later referencing it in a `SELECT` statement will trigger a L027 linting error, when more than one table is involved in the `FROM` clause.

### Expected Behaviour

As a user I would expect L027 to understand that I'm trying to use a variable declared earlier rather than a column alias. Similar to how L027 not fires in the USING case.

### Observed Behaviour

```
Unqualified reference 'test' found in select with more than one referenced table/view.
```

### How to reproduce

```sql
DECLARE test INT64 DEFAULT 42;

SELECT test
FROM t_table1
LEFT JOIN t_table_2
    ON TRUE
```

yields 

```
Unqualified reference 'test' found in select with more than one referenced table/view.
```

### Dialect

BigQuery

### Version

```
==== sqlfluff ====
sqlfluff:               0.11.2 python:                 3.10.3
implementation:        cpython dialect:                  bigquery
verbosity:                   2 templater:               jinja
```

### Configuration

N/A

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by tunetheweb** [comment]
Is it possible `test` is both a variable and a column reference in above example? If so, then the error could be valid, and SQLFluff has no way of knowing this.

Though guess we could make the assumption that it's the variable in these cases.

**Comment by barrywhart** [comment]
Seems like the potential impact of this goes far beyond any single rule. It basically suggests that for BigQuery (and potentially other dialects that support similar behavior), we'd need to add a step that processes any declarations like this and makes them available later in the file. Does BigQuery allow conditionals? If so, then it seems possible a variable declaration may or may not be executed, and potentially it's impossible to know until the SQL is actually executed.

TL;DR: Potentially a significant amount of effort. Changing from "bug" to "enhancement".

**Comment by barrywhart** [comment]
Google has removed the documentation page for this feature. What does that mean?

<img width="1151" alt="image" src="https://user-images.githubusercontent.com/1678585/166160346-24523d2a-12ab-400f-83d0-7f2d243585c3.png">

**Comment by barrywhart** [comment]
@tunetheweb: To answer your earlier question about, what if it's a column as well as a variable, the Google documentation has an answer:
>If a variable and column share the same name, the column takes precedence.
>This returns column x + column x:

```
DECLARE x INT64 DEFAULT 0;
SET x = 10;

WITH Numbers AS (SELECT 50 AS x)
SELECT (x+x) AS result FROM Numbers;
```

I think we should close this issue as "won't fix". It requires way too much database information that SQLFluff doesn't have access to.

**Comment by tunetheweb** [comment]
On one hand I kind of agree.

On the other, I think SQLFluff should be conservative in how it applies rules and the lack of understanding variables could break SQL. Not so much for L027 (which is not `fix`-compatible) but would be for the similar L028 rule (which is `fix`-compatible).

Many SQL variants (all of them?) have the concept of variables and it could be useful for rules to be able to know those that are declared so rules can be a bit more liberal for those identifiers under the assumption they might be referring to the variables rather than the columns.

**Comment by mxcoder** [comment]
Hi, I'm just dealing with this issue too, as our team heavily relies on BigQuery scripting (`BEING-END`, top level `IF`, top level `ASSERT`, multiple variables being declared, etc) and indeed sqlfluff is raising exceptions here and there. I'll try to contribute back on a few things for the dialect, but what about setting some "opinionated" patterns for variable names, then anyone using sqlfluff + bigquery dialect **should** follow the naming convention.

Maybe that could help keep sqlfluff simple and functional.
